[16.0][IMP] tax shelter : UX improvements

victor-champonnois opened the pull request:
Issue: 
As soon as the declaration is computed, all lines of the declaration have a “draft” status. The non-eligible lines are in the same state as the others. The same happens upon validation. It is only when sending the certificate that the status of these lines becomes either "sent" or "not eligible". 

Proposal:
- compute the state no_eligible on creation of the certificates instances. 
- mutes the certificate that are not eligible in the declaration form view

internal task : https://gestion.coopiteasy.be/web#action=475&active_id=524&model=project.task&view_type=kanban&menu_id=536
@huguesdk

huguesdk commented on l10n_be_cooperator/models/tax_shelter_certificate.py:
```diff
@@ -169,19 +169,13 @@ def send_certificates(self):
-                certificate.total_amount_eligible
-                + certificate.total_amount_eligible_previously_subscribed
-                > 0
-            ):
+            if certificate.state == "validated":
```
instead of an `if` inside a `for`, this can work too (and is more explicit):
```python
for certificate in self.filtered(lambda x: x.state == "validated"):
```

victor-champonnois replied:
done !

huguesdk commented on l10n_be_cooperator/models/tax_shelter_declaration.py:
```diff
@@ -155,7 +155,9 @@ def compute_declaration(self):
     def validate_declaration(self):
         self.ensure_one()
-        self.tax_shelter_certificates.write({"state": "validated"})
+        for certificate in self.tax_shelter_certificates:
+            if certificate.state == "draft":
+                certificate.write({"state": "validated"})
```
```suggestion
        self.tax_shelter_certificates.filtered(lambda x: x.state == "draft").write(
            {"state": "validated"}
        )
```

victor-champonnois replied:
done ! does this improve performances ?

huguesdk replied:
i’m not sure how odoo does it, but like this it might do the operation in only one sql update, while with a for loop it would be forced to do an update per record.

huguesdk reviewed (approved):
lgtm! adding a test would be great. some optional suggestions.

victor-champonnois commented:
@huguesdk thank you for the review
- I added your suggestions
- added a test
- renamed some buttons to clarify the sending process
- improved the interface of the declaration form. We let the user sets the date, and then the fiscal and declaration years are computed. Also set the date by default based on the previous year.

huguesdk commented on l10n_be_cooperator/models/tax_shelter_declaration.py:
```diff
@@ -163,10 +176,14 @@ def _compute_certificates(self, entries, partner_certificate):
             if entry.type == "subscription" and not excluded:
                 ongoing_capital_sub += entry.total_amount_line
 
+            certificate.compute_not_eligible()
```
since in this loop we are processing `subscription.register` entries, not certificates, doing this here might compute the status multiple times for the same certificate, and even result in a wrong status. this should go out of the loop and the function should iterate over all the certificates.

moreover, the `partner_certificate` function argument is useless (this is not something you touched). it can be initialized to an empty dict at the beginning of the function.

huguesdk commented on l10n_be_cooperator/models/tax_shelter_certificate.py:
```diff
@@ -62,7 +62,7 @@ class TaxShelterCertificate(models.Model):
             ("draft", "Draft"),
             ("validated", "Validated"),
-            ("no_eligible", "Not Eligible"),
+            ("not_eligible", "Not Eligible"),
```
don’t forget the migration script to convert current values.

huguesdk commented on l10n_be_cooperator/views/tax_shelter_declaration_view.xml:
```diff
@@ -47,21 +47,21 @@ SPDX-License-Identifier: AGPL-3.0-or-later
+                            <field name="fiscal_year" widget="char" />
                             <field name="name" />
-                            <field name="fiscal_year" />
+                        </group>
+                        <group>
                             <field name="tax_shelter_type" />
                             <field name="tax_shelter_capital_limit" />
                             <field name="previously_subscribed_capital" />
                         </group>
-                        <group>
-                            <field name="date_from" />
-                            <field name="date_to" />
-                        </group>
```
i’m not sure about this: i think that it should be possible to select the year first, which should set the date range, which can be adjusted afterwards. what rationale did you follow? the fact that a declaration is always done for the current year, except when testing?

moreover, there is no test that `date_to` is after `date_from`. this should be added to avoid mistakes.

victor-champonnois replied:
>i’m not sure about this: i think that it should be possible to select the year first, which should set the date range, which can be adjusted afterwards.


I thought about this solution first, but the implemention is a bit weird. Fiscal Year is an integer, so we can select any integer, which can lead to very weird dates. 

>what rationale did you follow? the fact that a declaration is always done for the current year, except when testing?

I think selecting date is a much more standard way of preparing fiscal reports in Odoo. 
Also, since in some cases, the date should not cover the whole year, selecting dates forces the user to think about this case and avoid some errors.

victor-champonnois replied:
>moreover, there is no test that date_to is after date_from. this should be added to avoid mistakes.

There is a check that throws an error, but I need to add automatic tests, you're right

victor-champonnois commented:
@huguesdk all changes done !

huguesdk reviewed (approved):
all good, many thanks!

polchampion reviewed (commented):
Functional test ok!
Side issue : the "send again" button on declaration lines don't seem to trigger the sending of an email.

polchampion reviewed (approved):
successful functional test

huguesdk commented:
/ocabot merge nobump
